utils/s2n_map.c

```diff
@@ -246,3 +246,91 @@ S2N_RESULT s2n_map_free(struct s2n_map *map)
 
     return S2N_RESULT_OK;
 }
+
+S2N_RESULT s2n_map_size(struct s2n_map *map, uint32_t *size)
+{
+    RESULT_ENSURE_REF(map);
+    *size = map->size;
+    return S2N_RESULT_OK;
+}
+
+/* Update the internal state so that `current_index` will point to the next value
+ * in the table or will equal `map->capacity` if there are no more elements in
+ * the map. 
+ */
+S2N_RESULT s2n_map_iterator_advance(struct s2n_map_iterator *iter)
+{
+    RESULT_ENSURE_REF(iter);
+    RESULT_ENSURE_REF(iter->map);
+
+    while (++iter->current_index < iter->map->capacity) {
+        /* a value was found in the map */
+        if (iter->map->table[iter->current_index].key.size) {
+            return S2N_RESULT_OK;
+        }
+    }
+    /* no more values were found in the map */
+    return S2N_RESULT_OK;
```

Comment:
So the current behavior is that if you call s2n_map_iterator_next when at the end of the map, it continues to return the last value. Wouldn't a clear failure be less likely to result in tricky bugs?

Reply:
`s2n_map_advance` is just an internal helper method. The public `s2n_map_iterator_next` will error if you call it when there is no next element. This is covered in the test case at line `243`. Sorry that the diff for the test file is so hard to read :(